### PR TITLE
[PORT] Simple HD wallet

### PR DESCRIPTION
**Pull request by ptschip**
Here's the initial port which contains the basic wallet functionality.  There are many other commits that come later which are small enhancements which I think I should add in a second PR otherwise it might be difficult to review.

### CODE REVIEW

**Comment by ptschip** [comment]
I think I'll just add one more commit to this PR which is to add the HD/non-HD indicator on the  QT UI.

**Comment by ptschip** [comment]
The basic HD wallet now works!  Give it a try.  I'll continue working on a followup PR which mostly fixes minor issues.

One question though, the HD wallet is on by default in this PR, should we rather leave it off by default?

**Comment by ptschip** [comment]
The way the functionality works now around -usehd, is that we won't error out on init if the user is just upgrading their BU node and has an old style wallet.  The only time we error is when the user specifically launches with -usehd=<bool>...This way we can still default new wallets to HD but not error on old type wallets which would be a bit of an upgrade mess.

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -1000,30 +1003,59 @@ bool BitcoinGUI::handlePaymentRequest(const SendCoinsRecipient &recipient)
...
         break;
     case WalletModel::Unlocked:
-        labelEncryptionIcon->show();
-        labelEncryptionIcon->setPixmap(
+        /*
+                labelEncryptionIcon->show();
+                labelEncryptionIcon->setPixmap(
+                    platformStyle->SingleColorIcon(":/icons/lock_open").pixmap(STATUSBAR_ICONSIZE, STATUSBAR_ICONSIZE));
+                labelEncryptionIcon->setToolTip(tr("Wallet is <b>encrypted</b> and currently <b>unlocked</b>"));
+        */
```
Comment (gandrewstone) [inline src/qt/bitcoingui.cpp]: looks like this comment can just be deleted (same as the code)

File: src/qt/bitcoingui.cpp [hunk 2]
```diff
@@ -1000,30 +1003,59 @@ bool BitcoinGUI::handlePaymentRequest(const SendCoinsRecipient &recipient)
...
     case WalletModel::Locked:
-        labelEncryptionIcon->show();
-        labelEncryptionIcon->setPixmap(
+        /*
+                labelEncryptionIcon->show();
+                labelEncryptionIcon->setPixmap(
+                    platformStyle->SingleColorIcon(":/icons/lock_closed").pixmap(STATUSBAR_ICONSIZE,
+           STATUSBAR_ICONSIZE));
+                labelEncryptionIcon->setToolTip(tr("Wallet is <b>encrypted</b> and currently <b>locked</b>"));
+        */
```
Comment (gandrewstone) [inline src/qt/bitcoingui.cpp]: remove comment

File: src/qt/walletmodel.cpp [hunk 3]
```diff
@@ -642,3 +642,5 @@ bool WalletModel::saveReceiveRequest(const std::string &sAddress, const int64_t
     else
         return wallet->AddDestData(dest, key, sRequest);
 }
+
+bool WalletModel::hdEnabled() const { return wallet->IsHDEnabled(); }
```
Comment (gandrewstone) [inline src/qt/walletmodel.cpp]: add newline end of file

File: qa/rpc-tests/wallet-hd.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        non_hd_add = self.nodes[0].getnewaddress()
+        self.nodes[1].importprivkey(self.nodes[0].dumpprivkey(non_hd_add))
+
+        # This should be enough to keep the master key and the non-HD key 
+        self.nodes[1].backupwallet(tmpdir + "hd.bak")
+        #self.nodes[1].dumpwallet(tmpdir + "hd.dump")
+
+        # Derive some HD addresses and remember the last
+        # Also send funds to each add
+        print("Derive HD addresses ...")
```
Comment (gandrewstone) [inline qa/rpc-tests/wallet-hd.py]: use logging rather than print (see other rpc tests for an example)

File: qa/rpc-tests/wallet-hd.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            hd_add = self.nodes[1].getnewaddress()
+            self.nodes[0].sendtoaddress(hd_add, 1)
+            self.nodes[0].generate(1)
+        self.nodes[0].sendtoaddress(non_hd_add, 1)
+        self.nodes[0].generate(1)
+
+        self.sync_all()
+        assert_equal(self.nodes[1].getbalance(), num_hd_adds + 1)
+
+        print("Restore backup ...")
```
Comment (gandrewstone) [inline qa/rpc-tests/wallet-hd.py]: logging (and other places)

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (ptschip) [reply src/qt/bitcoingui.cpp]: done

File: src/qt/bitcoingui.cpp (hunk 2 above)
Comment (ptschip) [reply src/qt/bitcoingui.cpp]: done

File: src/qt/walletmodel.cpp (hunk 3 above)
Comment (ptschip) [reply src/qt/walletmodel.cpp]: done

File: qa/rpc-tests/wallet-hd.py (hunk 4 above)
Comment (ptschip) [reply qa/rpc-tests/wallet-hd.py]: done

File: qa/rpc-tests/wallet-hd.py (hunk 5 above)
Comment (ptschip) [reply qa/rpc-tests/wallet-hd.py]: done

**Comment by ptschip** [comment]
Last two commits are updates as requested by comments and then did a full rebase on current dev

**Pull request merged**